src/mbgl/mtl/buffer_resource.cpp

```diff
@@ -90,13 +95,42 @@ void BufferResource::update(const void* data, std::size_t updateSize, std::size_
     updateSize = std::min(updateSize, size - offset);
 
     if (updateSize > 0) {
+        auto& stats = context.renderingStats();
         if (buffer) {
-            if (auto* content = static_cast<uint8_t*>(buffer->contents())) {
-                std::memcpy(content + offset, data, updateSize);
+            if (auto* const content = static_cast<uint8_t*>(buffer->contents())) {
+                // Until we can be sure that the buffer is not still in use to render the
+                // previous frame, replace it with a new buffer instead of updating it.
+                auto& device = context.getBackend().getDevice();
+
+                const bool updateIsEntireBuffer = (offset == 0 && updateSize == size);
+                const auto copySource = updateIsEntireBuffer ? data : content;
+                auto newBuffer = NS::TransferPtr(device->newBuffer(copySource, size, usage));
```

Comment:
Would it be any more efficient to keep a set number of buffers alive in a ring buffer, instead of allocating and releasing buffers all the time?